### DISCUSSION TITLE
Should there be an end date to open ended streams?

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/flow#40] [Ideas]
@razgraf shared the following idea in Slack:

> Another thing that will be challenging will be the mental-model around end-dates and possible end-dated. E.g. let's say I pay an employee, but then they just announced they're leaving the company next month. I still wanna pay their salary, but on the 1st of next month I want that stream to stop

Currently, this behaviour can be achieved by the following (cc @andreivladbrg):
1. Sender only deposits the necessary withdrawable amount
2. Sender calls `cancel` after 1 month which sets `rps` to zero. In case sender doesn't call the `cancel`, the stream will continue to accumulate debt but recipient wouldn't be able to withdraw anything.

However, @razgraf pointed out that this might be a visual/cognitive blocker for users.

I have two comments on this:
- Having an end date wouldn't do justice to "open ended" name (just saying).
- With an end date, it can possibly compete Lockup Linear streams. The only difference between the two would be the flexibility for sender to deposit funds into stream anytime they want and I think OE may have an advantage over LL streams.

Other ideas to implement this feature:

1. This can be automated with tools like gelato since `cancel` function can be called anytime by the sender (cc @andreivladbrg)
2. We can implement execution via signatures in `cancel` that means sender can submit a signed data through the UI which can be used to execute cancel on their behalf but only after a specified date.

**Comment by razgraf** [comment]
Oh thank you Shub! Was just writing a discussion myself on this. Pasting here.

> While the debt system allows for a stream to stop its funding requirements and simply "exist", I believe we should also provide a way for people to choose an "end date". 
>
> The use-case for payroll is: imagine you have an employee you've been paying with Sablier for 2 years, and they announce they're leaving next month. You'd want the payroll stream to work until the 1st of the next month, but then it should stop. Options:
> 1. You do nothing, and start accruing debt - some sort of a soft-close. There's also no way (I think) for a UI to understand that stream has been softly terminated, so it will simply keep existing in your dashboard.
> 2. You setup this "end date" which sets the rate-per-second to zero on a given date.
> 
> This end date could be editable (so you can extend it ... maybe only before it's reached).
> 
> ---
> 
> The advantage of having a clear "close" (through a close-stream action, as well as a "set end date in the future" action) is that it lessens the burden of having to time your cancellation, as well as the cognitive load of accrued debt.

---

On the OP:

> Having an end date wouldn't do justice to "open ended" name (just saying).

Yes, but we can rebrand or still keep open-ended as a core feature, while offering this end-date as an advanced setting. It provides flexibility.

> With an end date, it can possibly compete Lockup Linear streams.

I agree. But is that a bad thing?

> This can be automated with tools like gelato 

Yup. Although having it inside the core means one less dependency and less coordination cost with an external tool (who pays for the gas, what happens if the tx is not included in a block etc.).

**Reply by smol-ninja** [reply]
> But is that a bad thing?

Not at all.

I am in favour of this proposal of having an optional end date for open-ended streams. It will make core contracts more flexible and open for experimentation with new use cases that we have not anticipated yet.

**Reply by smol-ninja** [reply]
In the current open ended design, with no definition of end date, streams are always cancelable. By having an end date, do you think we would need them to be non-cancelable as well?

**Reply by razgraf** [reply]
That's a good question - it would give us an extra-feature but I'm not sure non-cancelability as a principle works when there are no guarantees about the resources a stream runs on.

**Comment by smol-ninja** [comment]
Peteris asked to have an end date: https://github.com/sablier-labs/v2-open-ended/discussions/67. This also brings another question: Should there be a start date to open-ended streams?

**Reply by andreivladbrg** [reply]
an idea came to mind now, what if we calculate the end date based on the `balance` and the `rps`?

```solidity
   function getEndDate(uint256 streamId) public view returns (uint256) {
       return (block.timestamp + (getBalance(streamId) / getRatePerSecond(streamId)));
   }
```

**Reply by smol-ninja** [reply]
That seems more like a timestamp when the stream starts to accumulate debt. Quite useful for the UI dashboard. But my understanding is that Peteris has asked for an end date as in the timestamp after the stream stops updating states. We should discuss your approach with him during the call.

PS: what should we call the timestamp when the current balance exhausts and the stream starts to accumulate debt? Is calling it `endDate` self-explanatory?

**Reply by andreivladbrg** [reply]
have not looked at what Peteris said yet, will do

> timestamp after the stream stops updating states

states are updated only on actions made on the stream, not on specific timestamps

> We should definitely discuss your approach with him during the call

yes

**Reply by PaulRBerg** [reply]
Worth noting that Peteris said on the call that he would be cool with this purely informational method.

My two cents - 'end date' would not be a good name because:

1. It can be confused for an actual, literal end time whereby the stream would **stop** accumulating debt
2. It can be confused with the end time in Lockup
3. We should use the term 'time' instead of 'date' to reflect the underlying Unix timestamp

Some ideas:

- insolvency time
- termination time
- terminal time
- terminus
- drain point
- depletion time
- payout time

**Reply by andreivladbrg** [reply]
great suggestion, i like "insolvency time"

**Reply by smol-ninja** [reply]
> It can be confused for an actual, literal end time whereby the stream would stop accumulating debt

Thats what end time is supposed to be. End time is a time when the stream stops accumulating debt. I do agree with your point that it can create confusion. 

Among your ideas, I like the ones that contain "terminate" as it, technically speaking, terminates the stream. @andreivladbrg [as per ChatGPT](https://chat.openai.com/share/a389ba4b-a857-4a7f-a76c-f946bfe1e47f), insolvency time is the time when a person can no longer pay his debt i.e. in this context, it would not be appropriate as sender can still pay the debt. It is more appropriate when sender / recipient makes the stream void because they can no longer pay each other. If you agree, I am happy to call it "termination time".

**Reply by PaulRBerg** [reply]
See my latest [comment](https://github.com/sablier-labs/v2-open-ended/discussions/40#discussioncomment-9394397) below. `depletion time` has grown on me, and I think it's clearer than `termination time`.

**Reply by PaulRBerg** [reply]
> End time is a time when the stream stops accumulating debt

Isn't it more than that? An actual end time would _either_ stop accumulating debt _or_ stop making payments despite there being a deposit left.

**Reply by smol-ninja** [reply]
What do you mean by "stop making payments"? Does it mean the recipient won't be able to withdraw the already streamed amount? Or does it mean that the stream will stop streaming assets to the recipient? 

If it's the latter that you meant, then the way debt stops accruing is by setting `ratePerSecond` to zero. This also implies that stream stops streaming assets to recipients. That has been my understanding of the end time so far.

PS: I replied on the comment below.

**Comment by smol-ninja** [comment]
Should we implement the optional end date to open-ended-streams in the first release? We had Peteris asked about it specifically.

**Reply by andreivladbrg** [reply]
didn't he say that is not mandatory?

**Reply by andreivladbrg** [reply]
i would suggest to "test the waters" without an end date (in the contract storage)

we still can have an end date field in the UI, right @razgraf ?

**Reply by razgraf** [reply]
Yes we can, but no clear `stop` action on a stream would make it extremely hard to identify which streams were:
- pseudo-stopped by not funding them (as intended)
- forgotten and accruing debt

I'd say we should try our best to push streams into an end-state where it's clear they were canceled/stopped as intended. Including an end-date helps with this, so it would be amazing if you guys could include it in the first version.

**Reply by andreivladbrg** [reply]
Some questions:
- Should all streams have an end date? Isn't that contradictory to the concept of an 'open-ended' contract?
- If not all streams have an end date, could this cause a psychological challenge for the creator in understanding whether there is an end date?
- IMO the implementation should mirror the `expiration` from Merkle contracts: if an end date is set, consider it; otherwise, set it to zero and ignore it.

My fear is that this end time/date will make things more complicated for the user to understand

**Reply by razgraf** [reply]
Yes, your last question describes perfectly what this end date should behave like. It's an optional hard stop.

The fact that it does bring it closer to the concept of a "lockup" contract is again fair, although I'm still on the fence if it's a bad thing or a good thing. @PaulRBerg [preempted](https://github.com/sablier-labs/v2-open-ended/discussions/40#discussioncomment-9381771) here that, in order to reduce lack of clarity we could just refer to it as a hard "date when the rate resets to zero" instead of "the end date".

> My fear is that this end time/date will make things more complicated for the user to understand

Naturally. But we can always push it under an "advanced features" menu.

**Reply by andreivladbrg** [reply]
OK, thank you, got it.

**Reply by smol-ninja** [reply]
Good points.

> didn't he say that is not mandatory?

I think it was dicey. I remember him mentioning that OZ would have a contract with them where the payment would be for 1 year. So imo, most of the open-ended streams will still have a "softly agreed" end time instead of running forever. it is also useful to keep track of which streams will end at what time in the future which I suppose they all will.

> Should all streams have an end date? Isn't that contradictory to the concept of an 'open-ended' contract?

To your argument, I totally agree that it is contradictory to the name itself. But maybe we should change the name to convey it in a better way 😅. Like payroll/pay, you mentioned that it was payroll originally but we changed it because of some legal issues.

By including an optional end time, it might also confuse users with lockup but the main advantage of OE over Lockup is that it does not require upfront capital.

**Reply by andreivladbrg** [reply]
> Like payroll/pay, you mentioned that it was payroll originally but we changed it because of some legal issues.

@PaulRBerg is it still the case?

**Reply by PaulRBerg** [reply]
Yeah, we can't go with `Payroll`. At one point, a blockchain analytics company had thought that Sablier was an employer of all of our Sablier V1.0 users. Recall that in V1.0, we had a contract called [`Payrol.sol`](https://docs.sablier.com/contracts/v1/deployments#sablier-v10).

**Reply by smol-ninja** [reply]
Thats very funny of them to assume that.

**Comment by PaulRBerg** [comment]
I've caught up on all messages in this thread. My thoughts:

There seem to be two ideas floating around: (i) implementing an actual end time and (ii) adding an informational getter that predicts the Unix timestamp when the sender's balance will be depleted to zero.

Regarding the 2nd idea, I think we all agree that the getter would be helpful. Naming-wise, I propose we go with `depletion time` because it is the one that most accurately reflects what's going on, and it's sufficiently distinct from anything resembling `end time` (by comparison, `termination time` is semantically closer to that).

Now, onto the 1st idea. I'm sympathetic to it, although I'm inclined to implement if off-contract via a Gelato integration. Two reasons:

1. Adds complexity.
    - Recall that any additional bit of logic in a smart contract increases the complexity and the attack surface in a non-linear way.
    - The risk of a Gelato bot failing to execute a tx seems less important than a smart contract bug.
2. We wouldn't have to change the name.

However, for the sake of brainstorming, let's suppose we will proceed with implementing this idea. I do see the benefit of an easier and more tight-knit front-end implementation.

What would the new name of the contract be? Here are some suggestions:

- `Continuous`
- `Flow`
- `FlexFlow`
- `Everflow`
- `Progressive`
- `TopUp`

---

P.S. I am against a signature-based implementation (as suggested by @smol-ninja in the OP). It seems like the most complicated of the bunch: the complexity would increase both in the smart contracts and in the UI.

**Reply by razgraf** [reply]
Note: if we have a status for OE (like we do for Lockup) streams and we go with `depletion time`, the value should be in-sync with what we call [`Status.DEPLETED`](https://github.com/sablier-labs/v2-core/blob/9657fbd0a77821a088b34d460dd1117e5600bb43/src/types/DataTypes.sol#L67) for maximum clarity.

**Reply by PaulRBerg** [reply]
Good idea to have a status enum. Sounds like we should create an issue for it or a discussion in case @andreivladbrg and/or @smol-ninja disagree

**Reply by andreivladbrg** [reply]
Also recommended this here: https://github.com/sablier-labs/v2-open-ended/discussions/46#discussioncomment-9181598

For OE contracts, it would make more sense to have the enum actually stored in the contract, unlike in Lockup where we compute it dynamically.

**Reply by PaulRBerg** [reply]
@andreivladbrg not quite - if we implement this depletion time, the status would have to be computed dynamically.

**Reply by andreivladbrg** [reply]
> the status would have to be computed dynamically

can you please, provide more details why you say this? 

things work a little bit different compared with lockup contracts

**Reply by razgraf** [reply]
I think he means we have to perform a computation to check if the status is indeed `DEPLETED`. So there isn't really a way to store this statically (in a variable), it has to be dynamically computed on every call for status (with a `getStatus` method).

e.g. `if(current time >= depletion time) return Status.DEPLETED`

**Reply by andreivladbrg** [reply]
ok, I got it now,  thanks for explaining. I was a bit confused because I missed the part "and we go with depletion time". 
And PRB said, "although I'm inclined to implement if off-contract via a Gelato integration" and I agree with this.
 
> Note: if we have a status for OE (like we do for Lockup) streams and we go with depletion time, the value should be in-sync with what we call [Status.DEPLETED](https://github.com/sablier-labs/v2-core/blob/9657fbd0a77821a088b34d460dd1117e5600bb43/src/types/DataTypes.sol#L67) for maximum clarity.

It can't be the same as what we call `DEPLETED` in Lockup - i.e. "all assets have been withdrawn and/or refunded" 
For the OE case, `if (current time >= depletion time)` simply means that the stream won't have funds for streaming anymore, not that any action has been taken on the actual stream (such as withdrawing all assets or refunds)

--- 

also, if we don't add the variable `depletionTime` in storage then it would be calculated as : `depletionTime == block.timestamp + (getBalance(streamId) / getRatePerSecond(streamId))`

which means that we have the invariant that depletion time is actually greater than or equal to current time

**Reply by PaulRBerg** [reply]
> It can't be the same as what we call DEPLETED in Lockup

It absolutely can be the same. Depletion is what happens in both cases; it doesn't matter that in one case, users perform a manual withdrawal, and in the other, time itself passes. Funds are depleted.

>  if we don't add the variable `depletionTime` in storage 

That's not what I and @razgraf are talking about. We are explicitly saying that the status should be computed at runtime because it depends upon `block.timestamp`, which is a dynamic property.

**Reply by smol-ninja** [reply]
I just finished reading the conversation. Thanks for sharing your ideas @PaulRBerg and @razgraf. Depletion is easier to implement at runtime without adding any new storage. So that's OK.

1. **Stream status:** Am I correct when I say that the status changes to depletion as soon as the deposit balance equals withdrawable balance beyond which stream starts to accumulate debt?

```solidity
function getStreamStatus(uint256 streamId) external view returns (uint8) {
    // --- snip ---
    } else if (getBalance(streamId) - withdrawableAmountOf(streamId) == 0) {
        return Status.DEPLETED;
    }
    // --- snip ---
}
```

2. **Get depletion timestamp:** The timestamp when the stream will start to accumulate debt in the future. And what would it return after the stream has been depleted? `block.timestamp`?

```solidity
function getDepletionTimestamp(uint256 streamId) external view returns (uint8 ) {
    return block.timestamp + (getBalance(streamId) - withdrawableAmountOf(streamId)) / getRatePerSecond(streamId);
}
```
-------------

Now going back to the first idea: `endTime`, I don't think it adds much complexity.

Since streaming is a function of `ratePerSecond` so with `endTime`, the value of `ratePerSecond` will return only if `block.timestamp < endTime`, otherwise 0. Therefore I disagree that it adds complexity (please let me know if I failed to understand the complexity surface).

But as @andreivladbrg suggested at one point, we can just launch without `endTime` and in the future if we receive user requests for that, then we can consider adding it. Or as you said, gelato is always an option to alternatively implement `endTime` behaviour.

> We wouldn't have to change the name

This is a good point. @andreivladbrg has also raised it before and I do agree that having an end time would mean changing the name from `open-ended`. Choosing a new name is difficult too.

>  I am against a signature-based implementation (as suggested by @smol-ninja in the OP). It seems like the most complicated of the bunch: the complexity would increase both in the smart contracts and the UI.

Agree with you.

> What would the new name of the contract be?

If we are open to a new name, I like Sablier Flow (Sablier Lockup and Sablier Flow). I'd also love to hear others' suggestions on names.

**Reply by razgraf** [reply]
> changing the name from open-ended

I'd still argue for changing the name to something that sounds better, for marketing purposes. A single word would work best.

**Reply by PaulRBerg** [reply]
<img width="200" src="https://i.kym-cdn.com/photos/images/newsfeed/001/650/747/aaf.png"/>

> Am I correct when I say that the status changes to depletion as soon as the deposit balance equals withdrawable balance

Yes.

> what would it return after the stream has been depleted? `block.timestamp`?

Yes.

> we can just launch without endTime and in the future if we receive user requests for that, then we can consider adding

Yes.

> I'd still argue for changing the name to something that sounds better, for marketing purposes

Yes

**Reply by PaulRBerg** [reply]
Related discussion about changing the name: https://github.com/sablier-labs/v2-open-ended/discussions/83

**Reply by PaulRBerg** [reply]
> I disagree that it adds complexity (please let me know if I failed to understand the complexity surface).

Yeah, you may be right that it wouldn't add that much complexity. But let's just prioritize go-to-market and launch without `endTime` - if and when users continue requesting this feature, we will implement it in a future release. Or we will integrate with Gelato.

**Reply by smol-ninja** [reply]
Agree with you.

**Reply by smol-ninja** [reply]
Can you please confirm if you agree with the [below comment](https://github.com/sablier-labs/v2-open-ended/discussions/40#discussioncomment-9457391)? I will create an issue then, especially about `getDepletionTimestamp`.

**Reply by PaulRBerg** [reply]
Done ser

**Comment by smol-ninja** [comment]
Have we all agreed to the following?
- Include `getDepletionTimestamp` function to get time when the stream starts to accumulate debt.
- Do not include `endTime` i.e. a timestamp when the stream's `ratePerSecond` starts returning 0. Instead recommend using gelato to call `pause()` or `cancel()` on the stream at `endTime`

If not, could you please respond to my last comment?

@sablier-labs/solidity @razgraf.

**Reply by PaulRBerg** [reply]
can you please allow ~72 hours until I can catch up with the entire thread here

**Reply by smol-ninja** [reply]
Sure. I'd also advise you to read https://github.com/sablier-labs/v2-open-ended/discussions/87#discussioncomment-9493787 regarding `endTime`.

**Reply by PaulRBerg** [reply]
I agree with your proposed approach.

**Comment by smol-ninja** [comment]
Issue:
- https://github.com/sablier-labs/v2-open-ended/issues/89

**Comment by smol-ninja** [comment]
Arguments in favour of having an end date (for future release): [UX difficulties with unwanted debt #197](https://github.com/sablier-labs/flow/discussions/197).
